[DRAFT] feat: draft of lessons learned on publish security

lwasser opened the pull request:
This is the start of a draft post on publishing to pypi best practices. it should be super duper user friendly and easy to read. I will need feedback!!   

i will be working on some graphics and such next.

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
This is not tested and could be wrong.

lwasser replied:
oh, right, good point.  What is a good example of caching dependencies (other than what we say with ultranalytics)? I use caching when i'm publishing. 

we pip install hatch and use `hatch build`

What do you think about removing the example and just telling people that if they do cache dependencies for any reason, don't do it when the build is for publishing? 

for instance, doing it for a website build could be ok if it's a preview, etc. vs publishing a package?

mikehack999 replied:
Send me a message I can definitely help you recover back your account 👨‍💻

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
          python-version: '3.12'
```

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
```diff
@@ -0,0 +1,307 @@
+---
+layout: single
+title: "Staying Secure: Best Practices for Publishing Python Packages to PyPI"
+excerpt: "blah"
```
```suggestion
excerpt: "Learn how to secure your Python package PyPI publishing workflows and protect your package from attacks. This post covers actionable steps, using PyPI Trusted Publisher, and sanitizing workflows, to ensure your projects stay safe."
```

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
```diff
@@ -0,0 +1,307 @@
+---
+layout: single
+title: "Staying Secure: Best Practices for Publishing Python Packages to PyPI"
```
```suggestion
title: "How to Secure Your Python Packages on PyPI: Stop the Mining Madness"
```

lwasser commented:
hey @webknjaz @hugovk do you have a bit of time to review this post for accuracy? i'm trying to provide a more beginner-friendly overview of what we learned that our community can quickly understand and implement. 

We use the the workflow @webknjaz helped us create for the pyosmeta package. And we want to link to all of the pypi / psf posts that are relevant too. many thanks for your time!

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
Technically, the environment is attached to a job in a workflow. This is the information PyPI would receive when looking up the trust link. Environments, however, can also be additionally configured in repository settings on GitHub. The recommended environment name is `pypi`. And the recommended configuration is having required reviewers (which would pause said job on GH, requiring a human to click the button).

lwasser replied:
@webknjaz I have not set this up myself and just noticed this feature after reading your comment. thank you. I really don't think people know about this stuff. 

TODO: I'll add a small note about how to do that in the post with a screenshot of what the interface looks like.

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
IIRC @woodruffw was against claiming that TP is more secure, so we settled on saying that it's “as secure” but is more convenient in many regards in the README of `pypi-publish`.

lwasser replied:
oh. that's different from my understanding of trusted publishing. I thought that TP provides a secure "pipeline" between the workflow and PyPI vs. a token that I suppose provides access to PyPi, but it doesn't necessarily directly connect to the workflow (i.e., you can specify a specific environment and workflow file with a trusted publisher. 

This statement challenges my understanding of trusted publisher. @woodruffw IF you happen to have a moment to clarify that would be super helpful.

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
Personally, I like calling it a “trust link”. Perhaps, @woodruffw would have opinions on the wordings/metaphors here.

lwasser replied:
I like the pipe idea as it's a visual people can understand even if it's not perfect. An isolated pipe or channel where files (sdist, wheel) can "travel" and be verified between GitHub and PyPI. 🤷🏻‍♀️  maybe there is a better visual but i want to help people understand why this is a good way to design a build / publish workflow. 

I want to create a graphic to help people understand this, as I think some of these important concepts get lost because it's all so technical!

But i say that knowing that you and others are experts - i'm just trying to simplify so more people can apply some of these things!

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
I don't fully agree with that note on unauthorized publication. Unauthorized publication can be prevented by controlling the trigger of the release process. Or setting up required reviewers in the environment used in the publishing job (usually called `pypi`) — this would prevent said job from even starting until a trusted human being clicks a button on GH. Recording the environment name on the PyPI side, though, allows restricting the “identity” of the workflow+repo+env that is to be considered “trusted”. But this only works on the authentication level between PyPI and GH. But not between GH and human, which is what you also want to secure.

lwasser replied:
ok, I am still working through your comments, but after I plan to reread the entire thing, and i'll reframe it. concepts:

1. trusted publisher is a great way to go - not necessarily BETTER than tokens. Perhaps more streamlined and simpler to set up. it's think we just suggest people do that. it eliminates GitHub secretes AND it eliminates needing to set up a PyPI project FIRST using a user-scoped token (because the project doesn't exist on PyPI yet). 

People will forget the step of updating their token to be project-level scope. So a trusted publisher, in my mind, is less risky for a user because it's simpler.   

2. For your build workflows, use environments that are secured and controlled by trusted humans -- (I can provide screenshots of how to set that up). This means that a publish workflow won't run until someone with access to the repo physically OKs the build. 

maybe there is a graphic here to simplify and help people understand what you wrote above?

gh + human authorization
github + PyPI authorization 

These are two different but related and important concepts. And they are high-level. like if people need to remember 2 things - these are the things.

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
cc @woodruffw apparently we forgot to update the screenshots in the warehouse doc to show `pypi` as the environment name.

lwasser replied:
Oops! I can open an issue if that is helpful in warehouse?

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
Maybe, don't mention that the environment is optional if this post is focused on maximizing security.

lwasser replied:
good point!! changed to (they will see optional in the image!)

environment (STRONGLY recommended)

webknjaz commented on images/python-packaging/trusted-publisher-form.webp:
Please, replace this screenshot with one having `pypi` typed into the last field. `release` is misleading — we've updated it in many places in the PyPI docs already, and on the website form, in the placeholder. But it looks like I missed the screenshots.

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
> [!tip]
> *Pro tip:* consider updating the Bluesky username to `pyopensci.org`. This can be configured and verified in like 3 minutes (through either DNS or a `.well-known` document on the website, but adds credibility to the account). You can create a repository in the org, call in `.well-known`, and add a file with the account DID as follows: https://github.com/webknjaz/.well-known/commit/5a3c3104aca2cf8b55c54acc8cf3363692c4604d. Just don't forget to enable Pages in the repo settings before clicking `Verify`.

lwasser replied:
WOAH!! i never knew about this for bluesky!.  
thank you for this tip - and for all of the thoughtful feedback. i'm going to work on this some more tomorrow!

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
I started using the exact repository ID some time ago. This is more accurate and would survive renames and transfers:
```suggestion
    if: github.repository_id == 174412809
```
(looked up @ https://api.github.com/repos/pyOpenSci/pyopensci.github.io)

If you want to check for the org, you could use the org ID. Org names are mutable and if the org gets renamed for whatever reason (might be a user account that hosts a repo in some cases), this check would break.

lwasser replied:
Ok great. yes, I worry the ID would be a bit harder for users to understand/find/use, etc., and yes, readability too!! thank you! I'll resolve this.

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
When referencing files on GH UI, make sure to always replace the branch name (`main`) with a commit sha. This is because branches are moving pointers to different commits. When the file changes or gets moved, the unversioned links would be misleading or start pointing to some other line. This is the way to make permalinks that survive time.

lwasser replied:
fixed!

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
In the doc you already linked, there's an example for GitLab as well: https://docs.pypi.org/trusted-publishers/adding-a-publisher/#gitlab-cicd. There are 4 platforms that support this mechanism already, and I saw some Bitbucket folks asking around whether they can integrate.

lwasser replied:
fixed the breakout and added that link!! I somehow read that GitLab support wasn't available yet, but obviously, I wasn't looking in the right place! thanks!

webknjaz commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
3. Update your GitHub Actions workflow to reference the Trusted Publisher configuration.
```

A GitHub Action (singular) is a different thing — it's just one of the steps in a job that is reusable.

webknjaz reviewed (commented):
Here's some quick notes. I might've missed something.

Let's also ask @woodruffw for feedback.

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
OOOOH no way. i missed the Google Colab part!! that's a pretty bold move for a hacker? maybe they were new hackers ?

lwasser replied:
geez. it's amazing how quick and far-reaching this single incident was!

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
good point - fixed!!

- 🚫 Avoid risky trigger events in GitHub actions like `pull_request_target`
- Consider using publish workflows on GitHub that are triggered by a [GitHub release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) rather than a pull request merge.

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
ok. This is super helpful. I could remove the caching part altogether, given so many moving pieces and caching seems more nuanced then just locking down the workflow itself with better triggers and  tighter permissions. But I also have a question about the dependencies. Because it is my understanding and experience that pinning all deps can also be highly problematic for users. For instance, if you have an environment with many different packages and try installing a package with pinned dips, that could lead to significant user issues. 

Unless we're just talking about dips in the workflow itself, such as hatch, pypa/build, etc. 

web apps make sense to pin but i can't quite wrap my head around pinning deps for a package. Could you please help me understand? And do you agree - maybe for more beginner users we just drop caching altogether as a topic?

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
Great suggestions!! I may add these links lower down in the post at the end, so the takeaways are just text.. but I'll definitely include them. This is what I have now:

- If you reference branches in a pull request, clean or [sanitize branch names](https://docs.github.com/en/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names) in your workflow.
- Consider using [zizmor](https://woodruffw.github.io/zizmor/) to check your GitHub workflows for security vulnerabilities!

i haven't used zizmor yet but i saw a package that Hugo maintains uses it so i may try it out with stravalib and pyosmeta!!

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
Looking at the blog posts, it seemed like the hackers somehow got direct access to that repo. Did i read that wrong?

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
Ok, I hear you. I've added this:

- Delete old GitHub tokens that you are no longer using. And Refresh existing tokens that you need periodically. 

I'll likely remove 2FA altogether, given it's required. I doubt most users understand why making tokens infinitely valid is bad!

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
here, I'm referring to GitHub actions in general but our examples do use the pypa action . But you are making a good point. 

i'll come back to this comment and will make sure that point is super clear in the text.

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
i will add a link to that packaging guide. And i see why this example may direct users incorrectly. in fact, i remember you telling me this last year when we were working on pyosMeta to update that build!

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
Fixed!

lwasser commented:
NOTES: i'm doing a big rewrite (more of a reorg) of this content. I'll start with the high-level concepts of securing the environment / GitHub and GitHub-PyPI, and then the other things like sanitizing branch names and such can come next.
This gives users 2 key takeaways (plus adding Zizmor as a third!) to start with now rather than 10 things, which is a lot! so I'd avoid more reviews until I push these edits given the structure will change (to hopefully be more clear and more accurate!) 🚀 

i'm super appreciative of all of the feedback.

lwasser commented:
Ok i've made significant changes to the post and added a new figure that tries to show trusted publisher as a concept with the 2 step ci action part.

i've also added links throughout to pypa / pypi documentation. AND I've tested out gizmo!! it's great. i haven't tried as a pre-commit yet but plan to. 

![trusted-publisher-pypi-github](https://github.com/user-attachments/assets/8002dd8e-547d-4feb-9243-8da9745ab426)

Also @webknjaz that Bluesky pro tip was 🔥 !! I set that up quickly, and we now have a pyopensci.org domain. I added a nod to you at the top of the post for providing so much input!!

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
There are other things that we can learn too from the recent breach. Many of these will be identified if you set up zizmor. These are discussed below.
```

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
This is where I'm confused about pinning deps, as that doesn't seem like a good idea for users when I think about environments and dependency conflict nightmares.

hugovk replied:
For deploy workflows or anything creating release artifacts end users might use, I think I've heard from @sethmlarson that it makes sense to reduce the number of dependencies you're using where possible: you can't have problems with a hacked dependency you're not using.

That should also reduce the conflict nightmares when pinning things...

lwasser replied:
Thanks Hugo! This helps clarify what I was struggling with! I'll rewrite this sentence to discuss the concept more broadly! I see that seth also gave this a thumbs up!

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
Add: a few sentences on 

delete old tokens and secrets containing pypi tokens... this is easy for people to do and good for them to know about.

sethmlarson commented on _posts/2024-12-13-python-packaging-security.md:
Maybe remove "short-lived" from this sentence? The underlying implementation of Trusted Publishers uses short-lived credentials, but the link itself is durable!

lwasser replied:
fixed!

sethmlarson commented on _posts/2024-12-13-python-packaging-security.md:
Yossarian -> William Woodruff

sethmlarson commented on _posts/2024-12-13-python-packaging-security.md:
I think that creating a GitHub Release doesn't require admin access, only write access?

lwasser replied:
Yup - you're right! i'll fix this

https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository

sethmlarson reviewed (commented):
Awesome blog post, thanks Leah! I left some comments:

lwasser commented:
@sethmlarson thank you so much for the review!! I appreciate it!  🚀 
I'll address your comments next!

lwasser commented:
@all-contributors please add @sethmlarson for review

hugovk commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
```yaml
```

hugovk commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
A trigger event in a GitHub Action is an event that sets off an action to run. For instance, you might have a trigger that runs a linter like Black or Ruff when a new pull request is opened.
```

hugovk commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
In the example GitHub Action `.yaml` file below, you see a `release` trigger defined. This tells the action to only trigger the workflow when you publish a release.
```

hugovk reviewed (commented):
Looking good, mostly nitpicks.

And good to see @sethmlarson here, I was going to summon him!

willingc commented:
I love seeing comments by @sethmlarson, @hugovk and @webknjaz. Thank you for being part of the pyOpenSci community and sharing your expertise. 🐍 🐍

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
A Trusted Publisher setup creates a secure link between PyPI and your repository.
```

thanks Seth! fixed.

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
thank you for this. There are a few folks here I don't know yet :) William is one of them. this addresses Seth's comment above as well.

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
```suggestion
  # By using release as a trigger, only GitHub users and actions with write access to make releases to our repo can trigger the push to PyPI
  release:
```

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
and I learned a bit of grammar here too. thank you!

lwasser commented on _posts/2024-12-13-python-packaging-security.md:
Oh! thanks! I've seen console used, but I was never sure about when to use it vs console. that makes a lot of sense!

lwasser commented:
@all-contributors please add @hugovk for code, review

lwasser commented:
Ok this is going live today and will go out on social next week 🚀 So sorry for the delays.
